### PR TITLE
Device linking gizmo

### PR DESCRIPTION
**Pull request by Bod9001**
### Changelog:
CL: [New] A new admin tool to link devices that use the multitool interface e.g APCs and ACUs

### CODE REVIEW

File: UnityProject/Assets/Scripts/Messages/Server/DeviceUpdateLinkerInfoMessage.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	{
+		public string ConnectedDevices;
+		public uint MasterID;
+	}
+
+	public override void Process(NetMessage msg)
+	{
+		var Slaves = JsonConvert.DeserializeObject<List<uint>>(msg.ConnectedDevices);
+		InGameDeviceLinker.Instance.MastersData.Clear();
+		var Master = msg.MasterID.NetworkIdentity().GetComponent<IMultitoolMasterable>();
```
Comment (MaxIsJoe) [inline UnityProject/Assets/Scripts/Messages/Server/DeviceUpdateLinkerInfoMessage.cs:21]: no fail safe on this?

File: UnityProject/Assets/Scripts/Objects/Cargo/ConveyorBelts/ConveyorBelt.cs [hunk 2]
```diff
@@ -285,8 +290,29 @@ private void ChangeDirection()
 		#region Multitool Interaction
 
 		public MultitoolConnectionType ConType => MultitoolConnectionType.Conveyor;
-		public bool MultiMaster => true;
-		int IMultitoolMasterable.MaxDistance => int.MaxValue;
+
+		public IMultitoolMasterable Master { get; set; }
```
Comment (MaxIsJoe) [inline UnityProject/Assets/Scripts/Objects/Cargo/ConveyorBelts/ConveyorBelt.cs:294]: i will eat your kidneys

PUT THE FIELDS AND PROPERTIES AT THE TOP

File: UnityProject/Assets/Scripts/Objects/Cargo/ConveyorBelts/ConveyorBeltSwitch.cs [hunk 3]
```diff
@@ -204,19 +207,13 @@ public enum SwitchState
...
-			List<ConveyorBelt> InnewConveyorBelts = new List<ConveyorBelt>();
-			foreach (var Conveyor in Imasters)
-			{
-				InnewConveyorBelts.Add(Conveyor as ConveyorBelt);
-			}
-			AddConveyorBelt(InnewConveyorBelts);
-		}
+		[field: SerializeField] public bool MultiMaster { get; set; } = true; //TODO
+		[field: SerializeField] public int MaxDistance { get; set; } = 30;
+		[field: SerializeField] public bool CanRelink { get; set; } = true;
```
Comment (MaxIsJoe) [inline UnityProject/Assets/Scripts/Objects/Cargo/ConveyorBelts/ConveyorBeltSwitch.cs:215]: https://github.com/unitystation/unitystation/assets/34368774/54682f9a-3ab7-4b2c-b69f-31f5869262f8

File: UnityProject/Assets/Scripts/Objects/Engineering/APC.cs [hunk 4]
```diff
@@ -448,7 +448,7 @@ private bool EmergencyState
 		public MultitoolConnectionType ConType => MultitoolConnectionType.APC;
 
 		[SerializeField]
-		private bool multiMaster = true;
+		private bool multiMaster = true; //TODO
```
Comment (MaxIsJoe) [inline UnityProject/Assets/Scripts/Objects/Engineering/APC.cs:451]: your todo is to move these at the top
not just this line, everything at the top as well, i saw properties hiding in the middle of functions

File: UnityProject/Assets/Scripts/UI/Systems/AdminTools/DevTools/InGameDeviceLinker/InGameDeviceLinker.cs [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+
+
+		foreach (var Device in MastersData[ShowFor.ConType][ShowFor])
+		{
+			if (Device is IMultitoolSlaveableMultiMaster MultimasterDevice) //TODO?
+			{
+				LinkGizmo[Device] = GameGizmomanager.AddNewLineStatic(Device.gameObject, Vector3.zero,
+					ShowFor.gameObject,
+					Vector3.zero, Color.blue);
+				// foreach (var Master in MultimasterDevice.Masters)
```
Comment (MaxIsJoe) [inline UnityProject/Assets/Scripts/UI/Systems/AdminTools/DevTools/InGameDeviceLinker/InGameDeviceLinker.cs]: remove unused code to please our overlord, codacy.

File: UnityProject/Assets/Scripts/UI/Systems/AdminTools/DevTools/InGameDeviceLinker/InGameDeviceLinker.cs [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+								(Slave as Component)?.gameObject);
+						}
+						else if (Slave.Master != cursorObject.SelectedMaster || Slave.Master == null)
+						{
+							Slave.SetMasterEditor(cursorObject.SelectedMaster);
+							LinkGizmo[Slave] = GameGizmomanager.AddNewLineStatic(Slave.gameObject, Vector3.zero,
+								cursorObject.SelectedMaster.gameObject,
+								Vector3.zero, Color.green);
+
+							if (CustomNetworkManager.IsServer)
```
Comment (MaxIsJoe) [inline UnityProject/Assets/Scripts/UI/Systems/AdminTools/DevTools/InGameDeviceLinker/InGameDeviceLinker.cs:298]: for the love of god stop nesting stuff like this
split these into functions and keep if statements reversed.

**Review by MaxIsJoe — CHANGES_REQUESTED** [review]

File: UnityProject/Assets/Scripts/Objects/Cargo/ConveyorBelts/ConveyorBelt.cs (hunk 2 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Objects/Cargo/ConveyorBelts/ConveyorBelt.cs:294]: it's grouped with the multitool interaction code

File: UnityProject/Assets/Scripts/Objects/Engineering/APC.cs (hunk 4 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Objects/Engineering/APC.cs:451]: yeah, multimaster need to be worked on

File: UnityProject/Assets/Scripts/Objects/Cargo/ConveyorBelts/ConveyorBelt.cs (hunk 2 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Objects/Cargo/ConveyorBelts/ConveyorBelt.cs:294]: put it at the top

File: UnityProject/Assets/Scripts/Messages/Server/DeviceUpdateLinkerInfoMessage.cs (hunk 1 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Messages/Server/DeviceUpdateLinkerInfoMessage.cs:21]: is a server message so, would be the server messing up if this was null

File: UnityProject/Assets/Scripts/Objects/Cargo/ConveyorBelts/ConveyorBelt.cs (hunk 2 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Objects/Cargo/ConveyorBelts/ConveyorBelt.cs:294]: I would have to get every single one, and I've messed around it way too much time to get syncvars to work

File: UnityProject/Assets/Scripts/Messages/Server/DeviceUpdateLinkerInfoMessage.cs (hunk 1 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Messages/Server/DeviceUpdateLinkerInfoMessage.cs:21]: there is no chance that you could send the wrong net id over to the server?

File: UnityProject/Assets/Scripts/Objects/Cargo/ConveyorBelts/ConveyorBelt.cs (hunk 2 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Objects/Cargo/ConveyorBelts/ConveyorBelt.cs:294]: I don't know what that has to do with organizing code, but if there is actual glue somewhere in our codebase that breaks these if they're not within an interface's region; then that's the only time I will let something like this slide.

Please keep all variables at the top of the class before any functions, it makes navigation easier and isn't messy.

File: UnityProject/Assets/Scripts/Messages/Server/DeviceUpdateLinkerInfoMessage.cs (hunk 1 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Messages/Server/DeviceUpdateLinkerInfoMessage.cs:21]: Would error on the request
```
		public override void Process(NetMessage msg)
		{
			if (IsFromAdmin() == false) return;

			if (msg.Slave is NetId.Empty or NetId.Invalid)
			{
				var Master = msg.Master.NetworkIdentity().GetComponent<IMultitoolMasterable>();
				InGameDeviceLinker.Instance.Refresh();
				var Object = SentByPlayer.GameObject;
				DeviceUpdateLinkerInfoMessage.SendTo(Object, InGameDeviceLinker.Instance.MastersData[msg.MultitoolConnectionType][Master], Master.gameObject);
				return;
			}
			```

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Messages/Server/DeviceUpdateLinkerInfoMessage.cs:21]: ah okay then, carry on

File: UnityProject/Assets/Scripts/Objects/Engineering/APC.cs (hunk 4 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Objects/Engineering/APC.cs:451]: that's not my point, my point is that there are variables in the middle of functions

File: UnityProject/Assets/Scripts/Objects/Cargo/ConveyorBelts/ConveyorBelt.cs (hunk 2 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Objects/Cargo/ConveyorBelts/ConveyorBelt.cs:294]: What should we do with code that's just being worked on Being added on?

**Comment by Bod9001** [comment]
does it still need changes, 
the tests are failing because it's complaining about fire base

**Pull request merged**